framework/Models/ExternalModel.py

```diff
@@ -222,54 +222,63 @@ def _externalRun(self, Input):
     additionalKeys = []
     if '_indexMap' in Input.keys():
       additionalKeys.append('_indexMap')
     for key in Input.keys():
       if key in additionalKeys:
         modelVariableValues[key] = copy.copy(Input[key])
+    # make sure the instance we pass to the user has the variables it needs
     for key in list(self.modelVariableType.keys()) + additionalKeys:
-      # add the variable as a member of "self"
-      try:
-        CustomCommandExecuter.execCommand('self.'+ key +' = copy.copy(object["'+key+'"])',self=externalSelf,object=modelVariableValues) #exec('externalSelf.'+ key +' = copy.copy(modelVariableValues[key])')  #self.__uploadSolution()
-      # if variable name is too strange to be a member of "self", then skip it
-      except SyntaxError:
-        self.raiseAWarning('Variable "{}" could not be added to "self" due to complex name.  Find it in "Inputs" dictionary instead.'.format(key))
-    #else:
-    #  InputDict = Input
+      # preference is to make it a member of the insantiated object
+      if key.isidentifier():
+        setattr(externalSelf, key, copy.copy(modelVariableValues[key]))
+      # otherwise, just have it available in the dictionary instead. This feels redundant.
+      else:
+        self.raiseAWarning(f'Variable "{key}" could not be added to ExternalModel first argument due to complex name. ' +
+                            'Find it in the second (dictionary) argument instead.')
     # only pass the variables and their values according to the model itself.
-    for key in Input.keys():
-      if key in self.modelVariableType.keys() or key in additionalKeys:
+    for key in Input:
+      if key in self.modelVariableType or key in additionalKeys:
         InputDict[key] = Input[key]
 
+    # run the user's model
     self.sim.run(externalSelf, InputDict)
 
+    # collect the user results
     for key in self.modelVariableType:
-      try:
-        # Note, the following string can't be converted using {} formatting, at least as far as I can tell.
-        CustomCommandExecuter.execCommand('object["'+key+'"]  = copy.copy(self.'+key+')', self=externalSelf,object=modelVariableValues) #exec('modelVariableValues[key]  = copy.copy(externalSelf.'+key+')') #self.__pointSolution()
-      except (SyntaxError, AttributeError):
-        self.raiseAWarning('Variable "{}" cannot be read from "self" due to complex name.  Retaining original value.'.format(key))
-    for key in self.initExtSelf.__dict__.keys():
-      # Note, the following string can't be converted using {} formatting, at least as far as I can tell.
-      CustomCommandExecuter.execCommand('self.' +key+' = copy.copy(object.'+key+')', self=self.initExtSelf, object=externalSelf) #exec('self.initExtSelf.' +key+' = copy.copy(externalSelf.'+key+')')
+      # preferentially obtain values from the instantiated object
+      if key.isidentifier(): # checks valid python variable name
+        modelVariableValues[key] = copy.copy(getattr(externalSelf, key))
```

Comment:
here we assume key is always available in externalSelf, is it True?

Reply:
I believe this has the same assumptions that were used in the CustomCommandExecuter.

Worst case, I think getattr will give you a None and it will raise an appropriate error later.

Reply:
I think it will raise an AttributeError, the original "try ... except ..." will catch the error. If it raised the AttributeError, the error message will not be easy to understand.

Reply:
sure; changed the if conditional to check existence as well. Honestly, this module really needs the rework we've talked about, but being back-compatible would be challenging.